### PR TITLE
o1vm/riscv32i: introduce instruction set

### PR DESCRIPTION
**Pull request by dannywillems**

### CODE REVIEW

File: o1vm/src/interpreters/riscv32i/interpreter.rs [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+    ShiftLeftLogicalImmediate,     // slli
+    ShiftRightLogicalImmediate,    // srli
+    ShiftRightArithmeticImmediate, // srai
+    SetLessThanImmediate,          // slti
+    SetLessThanImmediateUnsigned,  // sltiu
+
+    AddImmediate, // addi
+    AndImmediate, // andi
+    XorImmediate, // xori
+    OrImmediate,  // ori
```
Comment (svv232) [inline o1vm/src/interpreters/riscv32i/interpreter.rs:56]: also needs jalr

File: o1vm/src/interpreters/riscv32i/interpreter.rs [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+
+#[derive(
+    Debug, Clone, Copy, Eq, PartialEq, EnumCount, EnumIter, Default, Hash, Ord, PartialOrd,
+)]
+pub enum SBInstruction {
+    #[default]
+    BranchEq, // beq
+    BranchNeq,              // bne
+    BranchLessThan,         // blt
+    BranchGe,               // bge
```
Comment (svv232) [inline o1vm/src/interpreters/riscv32i/interpreter.rs:77]: should be BranchGreaterThan (I fixed this later down the stack)

File: o1vm/src/interpreters/riscv32i/interpreter.rs [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+    BranchLessThanUnsigned, // bltu
+    BranchGreaterThanEqual, // bgeu
```
Comment (svv232) [inline o1vm/src/interpreters/riscv32i/interpreter.rs:79]: should be BranchGreaterThanEqualUnsigned (This is also fixed later in my PRs)

File: o1vm/src/interpreters/riscv32i/interpreter.rs [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+    AddUpperImmediate, // auipc
+}
+
+#[derive(
+    Debug, Clone, Copy, Eq, PartialEq, EnumCount, EnumIter, Default, Hash, Ord, PartialOrd,
+)]
+pub enum UJInstruction {
+    #[default]
+    JumpAndLink, // jal
+    JumpAndLinkRegister, // jalr
```
Comment (svv232) [inline o1vm/src/interpreters/riscv32i/interpreter.rs:98]: move to itype

**Review by svv232 — APPROVED** [review]

**Comment by svv232** [comment]
I have fixed the issues in later prs, if this were not the case I would not have approved.

File: o1vm/src/interpreters/riscv32i/interpreter.rs [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+            UInstruction::AddUpperImmediate => write!(f, "auipc"),
+        }
+    }
+}
+
+impl std::fmt::Display for UJInstruction {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            UJInstruction::JumpAndLink => write!(f, "jal"),
+            UJInstruction::JumpAndLinkRegister => write!(f, "jalr"),
```
[the remark this answers is not in the document]
Comment (dannywillems) [reply o1vm/src/interpreters/riscv32i/interpreter.rs:242]: Thanks!

**Pull request merged**